### ISSUE TITLE
[BUG] Unicode filenames cause crash when writing to log file

### ISSUE THREAD

**Issue by codeceptsDE** [rdiff-backup/rdiff-backup#344]
## Bug summary

rdiff-backup crashes on files containing some rare unicode characters ('\u0308', '\u2560', and probably many more).

## Version, Python, Operating System

```
Using rdiff-backup version 2.0.0
        with cpython C:\Program Files\rdiff-backup-2.0.0\rdiff-backup.exe version 3.7.5
        on Windows-10-10.0.17763, fs encoding utf-8
```

```
Exception ''charmap' codec can't encode character '\u0308' in position 94: character maps to <undefined>' raised of class '<class 'UnicodeEncodeError'>':
  File "rdiff_backup\Main.py", line 390, in error_check_Main
  File "rdiff_backup\Main.py", line 412, in Main
  File "rdiff_backup\Main.py", line 348, in take_action
  File "rdiff_backup\Main.py", line 437, in Backup
  File "rdiff_backup\backup.py", line 39, in Mirror
  File "rdiff_backup\backup.py", line 268, in patch
  File "rdiff_backup\log.py", line 140, in __call__
  File "rdiff_backup\log.py", line 151, in log_to_file
  File "c:\users\travis\.pyenv\pyenv-win\versions\3.7.5\lib\encodings\cp1252.py", line 19, in encode

Traceback (most recent call last):
  File "rdiff-backup", line 32, in <module>
  File "rdiff_backup\Main.py", line 390, in error_check_Main
  File "rdiff_backup\Main.py", line 412, in Main
  File "rdiff_backup\Main.py", line 348, in take_action
  File "rdiff_backup\Main.py", line 437, in Backup
  File "rdiff_backup\backup.py", line 39, in Mirror
  File "rdiff_backup\backup.py", line 268, in patch
  File "rdiff_backup\log.py", line 140, in __call__
  File "rdiff_backup\log.py", line 151, in log_to_file
  File "c:\users\travis\.pyenv\pyenv-win\versions\3.7.5\lib\encodings\cp1252.py", line 19, in encode
UnicodeEncodeError: 'charmap' codec can't encode character '\u0308' in position 94: character maps to <undefined>
[1000] Failed to execute script rdiff-backup
```

Same with ```'\u2560'```

## rdiff-backup call

```
rdiff-backup -v5 --print-statistics <from> <to>
```

## What happened and what did you expect?

rdiff-backup terminated with a python exception when it reached a filename containing said rare unicode characters.
Ideally rdiff-backup wouldn't crash, or at least not make the user restart the whole backup either by waiting or by skipping affected files.

## More information

Admittedly, these are weird unicode characters to have in filenames, and they resulted from files where the filename was messed up. Nevertheless, it is possible to have such characters in filenames, and it's really unfortunate that rdiff-backup chokes on them, resulting in you having to re-run the entire backup.

The problem itself was fixable quite easily for me, by just checking the file path of the previous, successful file in the log, looking for weird a file and renaming it. 

Seems to be when writing the log file, though I can't verify if the file was copied correctly beforehand anymore. I suspect the issue is that python tries to write a utf-8 character as part of the file name to a log file encoded as ANSI / cp1252. Maybe it's enough to skip the file name in the log when it contains unmappable unicode. Or the backup.log could be written with utf-8 encoding.

**Edit:** Did some more testing:
Lowering log level to -v4 indeed "fixes" this. Some filenames to reproduce:
```
thisFileIsFine.txt
unicodeJam ┏( -_-)┛ ┗(-_- )┓.txt
```

<details>
  <summary>Some more verbose logs with rdiff-backup -v9 --print-statistics R:\evil R:\buTest:</summary>
  
  ```
2020-05-02 15:00:35.874491 +0200  <CLIENT-15252>  Using rdiff-backup version 2.0.0
2020-05-02 15:00:35.874491 +0200  <CLIENT-15252>        with cpython C:\Program Files\rdiff-backup-2.0.0\rdiff-backup.exe version 3.7.5
2020-05-02 15:00:35.876491 +0200  <CLIENT-15252>        on Windows-10-10.0.17763, fs encoding utf-8
2020-05-02 15:00:35.878490 +0200  <CLIENT-15252>  Making directory R:/buTest
2020-05-02 15:00:35.879489 +0200  <CLIENT-15252>  Making directory R:/buTest/rdiff-backup-data
2020-05-02 15:00:35.879489 +0200  <CLIENT-15252>  Hardlinks disabled by default on Windows
2020-05-02 15:00:35.879489 +0200  <CLIENT-15252>  Unable to import module (py)xattr.
Extended attributes not supported on filesystem at R:/evil
2020-05-02 15:00:35.880489 +0200  <CLIENT-15252>  Unable to import module posix1e from pylibacl package.
POSIX ACLs not supported on filesystem at R:/evil
2020-05-02 15:00:35.881489 +0200  <CLIENT-15252>  -----------------------------------------------------------------
Detected abilities for source (read only) file system:
  Access control lists                         Off
  Extended attributes                          Off
  Windows access control lists                 On
  Case sensitivity                             Off
  Escape DOS devices                           On
  Escape trailing spaces                       Off
  Mac OS X style resource forks                Off
  Mac OS X Finder information                  Off
-----------------------------------------------------------------
2020-05-02 15:00:35.882488 +0200  <CLIENT-15252>  Making directory R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0
2020-05-02 15:00:35.883491 +0200  <CLIENT-15252>  Touching R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/5-_ a.snapshot.gz
2020-05-02 15:00:35.883491 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/5-_ a.snapshot.gz
2020-05-02 15:00:35.883491 +0200  <CLIENT-15252>  Touching R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/uniá
2020-05-02 15:00:35.884488 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/uniá
2020-05-02 15:00:35.887487 +0200  <CLIENT-15252>  Touching R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/A
2020-05-02 15:00:35.887487 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/a
2020-05-02 15:00:35.897483 +0200  <CLIENT-15252>  Touching R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/foo
2020-05-02 15:00:35.898485 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/foo
2020-05-02 15:00:35.899483 +0200  <CLIENT-15252>  Making directory R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/hl
2020-05-02 15:00:35.899483 +0200  <CLIENT-15252>  Touching R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/hardlinked_file1
2020-05-02 15:00:35.899483 +0200  <CLIENT-15252>  Hard linking R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/hl/hardlinked_file2 to R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/hardlinked_file1
2020-05-02 15:00:35.900482 +0200  <CLIENT-15252>  Directories on file system at R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0 are not fsyncable.
Assuming it's unnecessary.
2020-05-02 15:00:35.900482 +0200  <CLIENT-15252>  Unable to import module (py)xattr.
Extended attributes not supported on filesystem at R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0
2020-05-02 15:00:35.903481 +0200  <CLIENT-15252>  Unable to import module posix1e from pylibacl package.
POSIX ACLs not supported on filesystem at R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0
2020-05-02 15:00:35.904481 +0200  <CLIENT-15252>  Touching R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/dir_inc_check
2020-05-02 15:00:35.905481 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/dir_inc_check
2020-05-02 15:00:35.905481 +0200  <CLIENT-15252>  Touching R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/regfile
2020-05-02 15:00:35.906301 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/regfile
2020-05-02 15:00:35.906301 +0200  <CLIENT-15252>  Touching R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/high_perms_file
2020-05-02 15:00:35.906301 +0200  <CLIENT-15252>  Touching R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/high_perms_dir
2020-05-02 15:00:35.909303 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/high_perms_file
2020-05-02 15:00:35.913300 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/high_perms_dir
2020-05-02 15:00:35.913300 +0200  <CLIENT-15252>  Touching R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/symlinked_file1
2020-05-02 15:00:35.914300 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/symlinked_file2
2020-05-02 15:00:35.914300 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/symlinked_file1
2020-05-02 15:00:35.915300 +0200  <CLIENT-15252>  Touching R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/foo
2020-05-02 15:00:35.915300 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0/foo
2020-05-02 15:00:35.915300 +0200  <CLIENT-15252>  Deleting R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0
2020-05-02 15:00:35.916300 +0200  <CLIENT-15252>  Removing directory R:/buTest/rdiff-backup-data/rdiff-backup.tmp.0
2020-05-02 15:00:35.917299 +0200  <CLIENT-15252>  -----------------------------------------------------------------
Detected abilities for destination (read/write) file system:
  Ownership changing                           Off
  Hard linking                                 On
  fsync() directories                          Off
  Directory inc permissions                    Off
  High-bit permissions                         On
  Symlink permissions                          Off
  Extended filenames                           On
  Windows reserved filenames                   On
  Access control lists                         Off
  Extended attributes                          Off
  Windows access control lists                 On
  Case sensitivity                             Off
  Escape DOS devices                           On
  Escape trailing spaces                       On
  Mac OS X style resource forks                Off
  Mac OS X Finder information                  Off
-----------------------------------------------------------------
2020-05-02 15:00:35.921707 +0200  <CLIENT-15252>  Backup: escape_dos_devices = 0
2020-05-02 15:00:35.921707 +0200  <CLIENT-15252>  Backup: escape_trailing_spaces = 1
2020-05-02 15:00:35.922707 +0200  <CLIENT-15252>  Enabled use_compatible_timestamps
2020-05-02 15:00:35.927705 +0200  <CLIENT-15252>  Making directory R:/buTest/rdiff-backup-data/increments
2020-05-02 15:00:35.927705 +0200  <CLIENT-15252>  Symbolic links excluded by default on Windows
2020-05-02 15:00:35.927705 +0200  <CLIENT-15252>  Starting mirror R:/evil to R:/buTest
2020-05-02 15:00:35.930706 +0200  <CLIENT-15252>  Processing changed file .
2020-05-02 15:00:35.932703 +0200  <CLIENT-15252>  Processing changed file thisFileIsFine.txt
2020-05-02 15:00:35.933703 +0200  <CLIENT-15252>  Regular copying (b'thisFileIsFine.txt',) to R:/buTest/rdiff-backup.tmp.1
2020-05-02 15:00:35.939700 +0200  <CLIENT-15252>  Writing file object to R:/buTest/rdiff-backup.tmp.1
2020-05-02 15:00:35.941700 +0200  <CLIENT-15252>  Copying attributes from (b'thisFileIsFine.txt',) to R:/buTest/rdiff-backup.tmp.1
2020-05-02 15:00:35.941700 +0200  <CLIENT-15252>  Setting time of R:/buTest/rdiff-backup.tmp.1 to 1588424202
2020-05-02 15:00:35.942700 +0200  <CLIENT-15252>  Renaming R:/buTest/rdiff-backup.tmp.1 to R:/buTest/thisFileIsFine.txt
2020-05-02 15:00:35.943700 +0200  <CLIENT-15252>  Exception ''charmap' codec can't encode character '\u250f' in position 85: character maps to <undefined>' raised of class '<class 'UnicodeEncodeError'>':
  File "rdiff_backup\Main.py", line 390, in error_check_Main
  File "rdiff_backup\Main.py", line 412, in Main
  File "rdiff_backup\Main.py", line 348, in take_action
  File "rdiff_backup\Main.py", line 437, in Backup
  File "rdiff_backup\backup.py", line 39, in Mirror
  File "rdiff_backup\backup.py", line 268, in patch
  File "rdiff_backup\log.py", line 140, in __call__
  File "rdiff_backup\log.py", line 151, in log_to_file
  File "c:\users\travis\.pyenv\pyenv-win\versions\3.7.5\lib\encodings\cp1252.py", line 19, in encode

Traceback (most recent call last):
  File "rdiff-backup", line 32, in <module>
  File "rdiff_backup\Main.py", line 390, in error_check_Main
  File "rdiff_backup\Main.py", line 412, in Main
  File "rdiff_backup\Main.py", line 348, in take_action
  File "rdiff_backup\Main.py", line 437, in Backup
  File "rdiff_backup\backup.py", line 39, in Mirror
  File "rdiff_backup\backup.py", line 268, in patch
  File "rdiff_backup\log.py", line 140, in __call__
  File "rdiff_backup\log.py", line 151, in log_to_file
  File "c:\users\travis\.pyenv\pyenv-win\versions\3.7.5\lib\encodings\cp1252.py", line 19, in encode
UnicodeEncodeError: 'charmap' codec can't encode character '\u250f' in position 85: character maps to <undefined>
[15252] Failed to execute script rdiff-backup
  ```
</details>

This happened _before_ the file was copied successfully, so the target directory only contains thisFileIsFine.txt, not the other one.

## Possible Solution
Have not tested, but I suspect this is the culprit:
https://github.com/rdiff-backup/rdiff-backup/blob/036a50183d04b863e303cb7caea24a0eac74d497/src/rdiff_backup/log.py#L85

From https://docs.python.org/3.7/library/functions.html#open :
_encoding is the name of the encoding used to decode or encode the file. This should only be used in text mode. The default encoding is platform dependent (whatever locale.getpreferredencoding() returns), but any text encoding supported by Python can be used. See the codecs module for the list of supported encodings._

So python defaults to the Windows-1252 character set because no explicit encoding is specified. Setting this to utf-8 should fix it, I think.

**Comment by ericzolf** [comment]
I'm a bit surprised because I thought that the "good old" days of "Windows encoding" and Windows was Unicode all over the place, but I might be wrong. Note that the "fs encoding" is UTF-8.
Perhaps it's more a matter of setting the `errors` parameter to something more relax than `strict`. Anyway doesn't sound too difficult, let me have a look.

**Comment by ericzolf** [comment]
Jup, what a mess:

```
>python
Python 3.7.7 (tags/v3.7.7:d7c567b08f, Mar 10 2020, 10:41:24) [MSC v.1900 64 bit (AMD64)] on win32 
Type "help", "copyright", "credits" or "license" for more information.
>>> import locale 
>>> locale.getpreferredencoding(False) 
'cp1252' 
```
And there is no clean way to change this globally... And `rpath.open` doesn't accept currently all the options of open, also because it does nifty things with compressing and/or remote writing/reading. So it won't be as easy as I thought...

**Comment by ericzolf** [comment]
OK, I took the "workaround" way, anything else would have been beyond a simple fix...

**Comment by ikus060** [comment]
Hello Eric, you might consider upgrading to py3.8. as I recall a lot of and
encoding stuff are now fixed in python 3.8. might be easier and cleaner to
simply upgrade python instead of implementing a workaround again for
Windows.

**Comment by ericzolf** [comment]
Sadly, it isn't possible: pyinstaller doesn't support it yet, see https://github.com/pyinstaller/pyinstaller/issues/4311 - they state that they need funding to implement it.

This said, the issue isn't related to the _filename_ encoding issues we encountered a lot already, but to the encoding of the file content itself (I'll comment a bit more in the PR), so it might not even help.
